src/ui/public/index_patterns/_index_pattern.js

```diff
@@ -13,318 +13,332 @@ import IndexPatternsFieldListProvider from 'ui/index_patterns/_field_list';
     title: 'string',
     timeFieldName: 'string',
     notExpandable: 'boolean',
     intervalName: 'string',
     fields: 'json',
     fieldFormatMap: {
       type: 'string',
-      _serialize: function (map) {
-        if (map == null) return;
-
-        let count = 0;
-        let serialized = _.transform(map, function (flat, format, field) {
-          if (!format) return;
-          count++;
-          flat[field] = format;
-        });
-
-        if (count) return angular.toJson(serialized);
+      _serialize(map = {}) {
+        const serialized = _.transform(map, serialize);
+        return angular.toJson(serialized);
       },
-      _deserialize: function (map) {
-        if (map == null) return {};
-        return _.mapValues(angular.fromJson(map), function (mapping) {
-          let FieldFormat = fieldformats.byId[mapping.id];
-          return FieldFormat && new FieldFormat(mapping.params);
-        });
+      _deserialize(map = '{}') {
+        return _.mapValues(angular.fromJson(map), deserialize);
       }
     }
   });
 
-  function IndexPattern(id) {
-    let self = this;
+  function serialize(flat, format, field) {
+    if (!format) {
+      return;
+    }
+    flat[field] = format;
+  }
+
+  function deserialize(mapping) {
+    const FieldFormat = fieldformats.byId[mapping.id];
+    return FieldFormat && new FieldFormat(mapping.params);
+  }
+
+  function updateFromElasticSearch(indexPattern, response) {
+    if (!response.found) {
+      throw new errors.SavedObjectNotFound(type, indexPattern.id);
+    }
 
-    setId(id);
+    _.forOwn(mapping, (fieldMapping, name) => {
+      if (!fieldMapping._deserialize) {
+        return;
+      }
+      response._source[name] = fieldMapping._deserialize(
+        response._source[name], response, name, fieldMapping
+      );
+    });
+
+    // give index pattern all of the values in _source
+    _.assign(indexPattern, response._source);
+
+    indexPattern._indexFields();
+
+    // any time index pattern in ES is updated, update index pattern object
+    indexPattern._docSource
+      .onUpdate()
+      .then(response => updateFromElasticSearch(indexPattern, response), notify.fatal);
+  }
 
-    let docSource = new DocSource();
+  class IndexPattern {
+    constructor(id) {
+      this._setId(id);
+      this._docSource = new DocSource();
+      this.metaFields = config.get('metaFields');
+      this.getComputedFields = getComputedFields.bind(this);
 
-    self.init = function () {
-      // tell the docSource where to find the doc
-      docSource
-      .index(kbnIndex)
-      .type(type)
-      .id(self.id);
+      this.flattenHit = flattenHit(this);
+      this.formatHit = formatHit(this, fieldformats.getDefaultInstance('string'));
+      this.formatField = this.formatHit.formatField;
+
+      this.routes = IndexPattern.routes;
+    }
+
+    _setId(id) {
+      this.id = id;
+      return id;
+    }
+
+    _watch() {
+      if (this._watching) {
```

Comment:
Is there a risk of this running multiple times? I usually prefer to wrap these types of things in `_.once()` unless it's necessary to revert the effect.